evm/src/recursive_verifier.rs

```diff
@@ -1,148 +1,574 @@
+            let proof = &self.recursive_proofs[i];
+            let pt = builder.add_virtual_proof_with_pis(&verifier_data.common);
+            pw.set_proof_with_pis_target(&pt, proof);
+            let inner_data = VerifierCircuitTarget {
+                constants_sigmas_cap: builder
+                    .add_virtual_cap(verifier_data.common.config.fri_config.cap_height),
+            };
+            pw.set_cap_target(
+                &inner_data.constants_sigmas_cap,
+                &verifier_data.verifier_only.constants_sigmas_cap,
+            );
+            (pt, inner_data)
+        });
+
+        let pis: [_; NUM_TABLES] = std::array::from_fn(|i| {
+            PublicInputs::from_vec(&proofs_target[i].0.public_inputs, inner_config)
+        });
+
+        let mut challenger = RecursiveChallenger::<F, C::Hasher, D>::new(builder);
+        for pi in &pis {
+            for h in &pi.trace_cap {
+                challenger.observe_elements(h);
+            }
+        }
+        let ctl_challenges = get_grand_product_challenge_set_target(
+            builder,
+            &mut challenger,
+            inner_config.num_challenges,
+        );
+        // Check that the correct CTL challenges are used in every proof.
+        for pi in &pis {
+            for i in 0..inner_config.num_challenges {
+                builder.connect(
+                    ctl_challenges.challenges[i].beta,
+                    pi.ctl_challenges.challenges[i].beta,
+                );
+                builder.connect(
+                    ctl_challenges.challenges[i].gamma,
+                    pi.ctl_challenges.challenges[i].gamma,
+                );
+            }
+        }
+
+        challenger.duplexing(builder);
+        let state = challenger.state();
+        for k in 0..SPONGE_WIDTH {
+            builder.connect(state[k], pis[0].challenger_state_before[k]);
+        }
+        // Check that the challenger state is consistent between proofs.
+        for i in 1..NUM_TABLES {
+            for k in 0..SPONGE_WIDTH {
+                builder.connect(
+                    pis[i].challenger_state_before[k],
+                    pis[i - 1].challenger_state_after[k],
+                );
+            }
+        }
+
+        // Verify the CTL checks.
+        let degrees_bits = std::array::from_fn(|i| verifier_data[i].common.degree_bits);
+        verify_cross_table_lookups_circuit::<F, C, D>(
+            builder,
+            self.cross_table_lookups,
+            pis.map(|p| p.ctl_zs_last),
+            degrees_bits,
+            ctl_challenges,
+            inner_config,
+        );
+        for (i, (proof_target, inner_data)) in proofs_target.into_iter().enumerate() {
+            builder.verify_proof(proof_target, &inner_data, &verifier_data[i].common);
+        }
+    }
+}
+
+/// Recursively verify a Stark proof.
+/// Outputs the recursive proof and the associated verifier data.
+fn recursively_verify_stark_proof<
```

Comment:
Here too, it seems like this method couples preprocessing with witness generation, but I think we should separate them.

Reply:
There's `verifier_data_recursive_stark_proof` that generates the verifier data without witness generation, is that what you mean?

Reply:
So if I understand right, this function does
- create virtual targets, mainly `add_virtual_stark_proof`
- circuit building, same as `verifier_data_recursive_stark_proof`
- witness generation, mainly `set_stark_proof_target`

For our production use case I think we'll want to separate those things, so I guess we'd be calling `verifier_data_recursive_stark_proof`, `set_stark_proof_target`, etc. separately, and wouldn't use this function.

If it's convenient for testing we could keep it, but if so I think we should have it call `verifier_data_recursive_stark_proof` to remove the duplication, and add `cfg(test)` so it's clear it doesn't need to be audited.

Reply:
Oh yeah good point, I moved them to a `mod tests` and we can remove them when we remove the `all_stark.rs` tests.